### PR TITLE
Support HDFS Tiering

### PR DESCRIPTION
**Pull request by kevcunnane**
This is the 1st step to supporting HDFS Tiering. Note: Icon needs updating, and I may need to get a linked file icon added too.

**Changes**:
- Add new mounted folder icon
- Disable delete/mkdir/upload for mounted files and folders
- Disable delete for root HDFS folder (this was added in error)

**Screenshots**:
Tree:
image
Context menu on mounted folder:
image
Context menu on mounted file:
image

### CODE REVIEW

File: extensions/mssql/src/hdfs/webhdfs.ts [hunk 1]
```diff
@@ -465,6 +466,27 @@ export class WebHDFS {
 		});
 	}
 
+	/**
+	 * Get ACL status for given path
+	 * @param path The path to the file/folder to get the status of
```
Comment (Charles-Gagnon) [inline extensions/mssql/src/hdfs/webhdfs.ts]: Update description and params

File: extensions/mssql/src/objectExplorerNodeProvider/hdfsProvider.ts [hunk 2]
```diff
@@ -161,12 +173,26 @@ export class FolderNode extends HdfsFileSourceNode {
...
+			iconType: this.isMounted() ? 'Folder_mounted' : 'Folder'
 		};
 		return nodeInfo;
 	}
 
+	private isMounted(): boolean {
+		return this.mountStatus === MountStatus.Mount || this.mountStatus === MountStatus.Mount_Child;
+	}
+
+	private getSubType(): string {
```
Comment (Charles-Gagnon) [inline extensions/mssql/src/objectExplorerNodeProvider/hdfsProvider.ts]: >string [](start = 23, length = 6)

nit: return string | undefined

**Comment by chlafreniere** [comment]
Do all of the actions work as expected after removing those that didn't make sense? Does Analyze in Notebook work?

File: extensions/mssql/src/hdfs/webhdfs.ts [hunk 3]
```diff
@@ -465,6 +466,27 @@ export class WebHDFS {
...
+	 * @param callback Callback to handle the response
+	 * @returns void
+	 */
+	public getMounts(callback: (error: HdfsError, mounts: Mount[]) => void): void {
+		let endpoint = this.getOperationEndpoint('listmounts', '');
+		this.sendRequest('GET', endpoint, undefined, (error, response) => {
+			if (!callback) { return; }
+			if (error) {
+				callback(error, undefined);
+			} else if (response.body.hasOwnProperty('Mounts')) {
```
Comment (chlafreniere) [inline extensions/mssql/src/hdfs/webhdfs.ts]: will it always be capitalized?

File: extensions/mssql/src/objectExplorerNodeProvider/fileSources.ts [hunk 4]
```diff
@@ -147,18 +150,43 @@ export class FileSourceFactory {
...
-	public enumerateFiles(path: string): Promise<IFile[]> {
+	public async enumerateFiles(path: string, refresh?: boolean): Promise<IFile[]> {
+		if (!this.mounts || refresh) {
+			await this.loadMounts();
+		}
+		return this.readdir(path);
+	}
+
+	private loadMounts(): Promise<void> {
+		return new Promise((resolve, reject) => {
```
Comment (chlafreniere) [inline extensions/mssql/src/objectExplorerNodeProvider/fileSources.ts]: With these changes, is our automation around promises being used correctly happy?

Not saying that I\'m spotting anything necessarily, just want to make sure that we're good there.

File: extensions/mssql/src/hdfs/webhdfs.ts (hunk 3 above)
Comment (kevcunnane) [reply extensions/mssql/src/hdfs/webhdfs.ts]: That's the official contract, yeah. I can try to check both but we're not doing this elsewhere

File: extensions/mssql/src/objectExplorerNodeProvider/fileSources.ts (hunk 4 above)
Comment (kevcunnane) [reply extensions/mssql/src/objectExplorerNodeProvider/fileSources.ts]: What does the automation do again? This is the same basic pattern as previous - we use Promises, and this is mostly converting callbacks into those?

**Comment by kevcunnane** [comment]
@chlafreniere All other actions make sense. Analyze in Notebook is a "read file" operation which does make sense. Similarly you can mount a folder and create external table over it, that's an important scenario. So the remainder are all valid.

File: extensions/mssql/src/objectExplorerNodeProvider/fileSources.ts (hunk 4 above)
Comment (kevcunnane) [reply extensions/mssql/src/objectExplorerNodeProvider/fileSources.ts]: Verified this shouldn't be impacted

**Pull request merged**